File: frontend/src/lib/audit.ts
"use client";

import { apiFetch } from "./api";

export type AuditLog = {
  id: string;
  action: string;
  category: string;
  target: string | null;
  ip_address: string | null;
  created_at: string;
};

/** List the authenticated user's audit logs, optionally filtered. */
export function listAuditLogs(
  token: string,
  opts: { category?: string; search?: string } = {},
) {
  const params = new URLSearchParams();
  if (opts.category && opts.category !== "all") params.set("category", opts.category);
  if (opts.search) params.set("search", opts.search);
  const qs = params.toString();
  return apiFetch<AuditLog[]>(`/v1/audit-logs${qs ? `?${qs}` : ""}`, { token });
}

File: crates/api/src/routes/audit.rs
//! Audit log listing.

use crate::auth::authenticate;
use crate::error::{ApiError, ApiResult, Envelope};
use crate::state::AppState;
use axum::extract::{Query, State};
use axum::http::HeaderMap;
use axum::Json;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct AuditQuery {
    /// Filter by category (authentication, wallet, address, credentials, configuration).
    pub category: Option<String>,
    /// Case-insensitive search over action/target.
    pub search: Option<String>,
}

/// `GET /v1/audit-logs` — the authenticated user's activity, newest first.
pub async fn list_audit_logs(
    State(state): State<AppState>,
    headers: HeaderMap,
    Query(q): Query<AuditQuery>,
) -> ApiResult<Json<Envelope<Vec<octo_store::AuditLog>>>> {
    let user_id = authenticate(&headers, &state)?;
    let category = q.category.filter(|c| !c.is_empty() && c != "all");
    let search = q.search.filter(|s| !s.is_empty());

    let rows = state
        .store()
        .list_audit_logs(user_id, category.as_deref(), search.as_deref(), 200)
        .await
        .map_err(|_| ApiError::Internal)?;
    Ok(Envelope::ok(rows))
}
